lib/datasets/mnist.rb

```diff
@@ -0,0 +1,96 @@
+require 'zlib'
+
+require_relative "dataset"
+
+
+module Datasets
+  class Mnist < Dataset
+
+    Record = Struct.new(:pixels, :label)
+
+    def initialize(type: :train)
+      unless [:train, :test].include?(type)
+        raise 'Please set type :train or :test'
+      end
+
+      super()
+
+      @metadata.name = "mnist-#{type}"
+      @metadata.url = "http://yann.lecun.com/exdb/mnist/"
+      @type = type
+
+      case type
+      when :train
+        @metadata.description = "a training set of 60,000 examples"
+      when :test
+        @metadata.description = "a test set of 10,000 examples"
+      end
+    end
+
+    def each(&block)
+      return to_enum(__method__) unless block_given?
+
+      image_path = cache_dir_path + target_file(:img)
+      label_path = cache_dir_path + target_file(:label)
+      base_url = "http://yann.lecun.com/exdb/mnist/"
+
+      unless image_path.exist?
+        download(image_path, base_url + target_file(:img))
+      end
+
+      unless label_path.exist?
+        download(label_path, base_url + target_file(:label))
+      end
+
+      open_data(image_path, label_path, &block)
+    end
+
+    private
+    def open_data(image_path, label_path, &block)
+      images = parse_images(image_path)
+      lables = parse_labels(label_path)
+      images.zip(lables) do |row|
```

Comment:
Can we implement stream parsing instead of "parsing then yielding records"?

Reply:
I implemented stream parsing to only `Image data` but not to `Label data`, because I think label data is not so big that it need stream parsing.
If you think it need  to `Label data` too, for reasons such as readability, efficiency etc... , I'll fixed it!

Reply:
I noticed it now, we may not able to implement stream parsing to `Label data` ...